presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SimplifyArrayOperationsRewriter.java

```diff
@@ -0,0 +1,246 @@
+
+    public static Expression simplifyArrayOperations(Expression expression, Session session)
+    {
+        if (SystemSessionProperties.isSimplifyArrayOperations(session)) {
+            return ExpressionTreeRewriter.rewriteWith(new Visitor(), expression);
+        }
+
+        return expression;
+    }
+
+    private static class Visitor
+            extends ExpressionRewriter<Void>
+    {
+        private static boolean isConstant(Expression expression)
+        {
+            if (expression instanceof Cast) {
+                return isConstant(((Cast) expression).getExpression());
+            }
+
+            return expression instanceof Literal;
+        }
+
+        private static boolean isZero(Expression expression)
+        {
+            if (expression instanceof Cast) {
+                return isZero(((Cast) expression).getExpression());
+            }
+
+            return expression instanceof LongLiteral && ((LongLiteral) expression).getValue() == 0;
+        }
+
+        private static boolean isCardinalityOfFilter(Expression expression)
+        {
+            if (expression instanceof FunctionCall) {
+                FunctionCall functionCall = (FunctionCall) expression;
+                return functionCall.getName().toString().equals("cardinality") &&
+                        functionCall.getArguments().size() == 1 &&
+                        functionCall.getArguments().get(0) instanceof FunctionCall &&
+                        ((FunctionCall) (functionCall.getArguments().get(0))).getName().toString().equals("filter");
+            }
+
+            return false;
+        }
+
+        private static boolean isSimpleComaparison(ComparisonExpression comparisonExpression)
+        {
+            switch (comparisonExpression.getOperator()) {
+                case EQUAL:
+                case GREATER_THAN:
+                case LESS_THAN:
+                case NOT_EQUAL:
+                case GREATER_THAN_OR_EQUAL:
+                case LESS_THAN_OR_EQUAL:
+                    return true;
+            }
+
+            return false;
+        }
+
+        private Expression simplifyCardinalityOfFilterComparedToZero(
+                NodeLocation location,
+                Expression arrayArg,
+                ComparisonExpression.Operator operator,
+                LambdaArgumentDeclaration origLambdaArgument,
+                Expression origLambdaBody, String newLambdaArgumentName,
+                String finalLambdaArgumentName,
+                LambdaArgumentDeclaration finalLambdaArgument)
+        {
+            // Comparing to zero can be rewrtitten to a boolean.
+            // cardinality(filter(a, x -> f(x)) = 0 becomes reduce(a, false, (s, x) -> case when s then s else f(x) end, s -> !s)
+            // cardinality(filter(a, x -> f(x)) > 0 becomes reduce(a, false, (s, x) -> case when s then s else f(x) end, s -> s)
+            Expression finalLambdaBody = new Identifier(location, finalLambdaArgumentName, false);
```

Comment:
Let's name this `outputFunction` to be consistent with documentation of reduce function.